### PR TITLE
Code review PR

### PR DESCRIPTION
**Pull request by myronmarston**
@TravisSpangle I've created this PR with your commits so I have something to review.

### CODE REVIEW

File: app/food_truck.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# Retrieves food-truck feed and parses it
+
+require_relative '../constants'
+require 'httparty'
+require 'nokogiri'
+
+module FoodTruck
+  class FoodTruck
```
Comment (myronmarston) [inline app/food_truck.rb:8]: IMO, it's confusing to name this `FoodTruck` given that the outer module is also `FoodTruck`.  In some contexts, it's not clear which constant you are referring to by `FoodTruck`.  For example, in your API:

``` ruby
module FoodTruck
  class API < Sinatra::Base
    get '/api' do
      ft = FoodTruck.new

      SpeakSlack.trucks(trucks: ft.trucks)
    end
  end
end
```

...it looks like `ft = FoodTruck.new` is trying to create an instance of the outer module, but of course that isn't possible, and looking at the entire code base I can see it is making a `FoodTruck::FoodTruck` instance.  But it's still confusing.

I recommend calling this something like `FoodTruck::ScheduleParser`.  That's it's purpose, right: parsing a published food truck schedule?

File: app/food_truck.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    attr_reader :trucks
+    def initialize(url: FOOD_TRUCK_URL)
+      load_trucks_from(url)
+    end
+
+    private
+
+    def load_trucks_from(url)
+      request = HTTParty.get(url)
```
Comment (myronmarston) [inline app/food_truck.rb:17]: The fact that you use `HTTParty` here to get the the schedule page means this class has two distinct responsibilities:
- Fetching the schedule page.
- Parsing the schedule page.

Abstractly, this is a violation of the [Single Responsibility Principle](https://blog.8thlight.com/uncle-bob/2014/05/08/SingleReponsibilityPrinciple.html). Concretely, it has some adverse effects:
- Tests for your parsing logic require an internet connection and are made slower by the need to fetch the markup they parse over the internet.
- Tests for your parsing logic are non-deterministic and unreliable.  They can fail any time due to a network problem or due to the remote server being down (for maintenance or whatever).  It's not good if you have tests that can randomly fail due to something you have no control over and that has nothing to do with the correctness of your logic here.
- Due to the non-determinism, your tests for this class make have very weak expectations: just that a non-empty array is returned.
- If there's some edge case bug that is occasionally hit in the parsing logic, there is no simple way to write a test for it because you can't deterministically feed it the markup that triggers the edge case in your test.
- If/when you add more robust logic around fetching the page (e.g. to handle transient network failures, or to handle redirects, or to use a caching proxy, or whatever) it's going to be entangled with your parsing logic here.

Instead, I would recommend you split this into two distinct classes:
- `FoodTruck::ScheduleFetcher`: responsible for fetching the schedule over the network.  (The HTTParty bits would go here, and this would ideally return a string of HTML markup).
- `FoodTruck::SchedulerParser`: responsible for parsing the schedule.  (This would receive the string of HTML markup a input).

This separation solves the problems above:
- It's easy to properly unit test this class with deterministic, robust tests: just create example markup strings and feed it in.
- Your expectations in those unit tests can be very string since the tests will be deterministic.
- Writing tests to cover edge cases that crop up becomes easy: just write a test that passes the problematic markup.
- You have distinct, non-entangled places to evolve the fetching and parsing logic individually.

If you go that route, you could either create a new class to encapsulate combining the fetcher and parser (basically, taking the place of this class) or you could combine them in the sinatra app.  I'd lean towards the latter -- the Sinatra app can coordinate these two things just fine, with no need for another layer of indirection, IMO.

File: app/speak_slack.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+# Takes an array of food truck names and formats it to slack
+require_relative '../constants'
+
+module FoodTruck
+  class SpeakSlack
+    def self.trucks(trucks: [])
```
Comment (myronmarston) [inline app/speak_slack.rb:6]: `trucks` doesn't communicate to me what this method is for.  It looks like maybe it's for formatting?  Maybe `FoodTruck::SlackFormatter.format(trucks)` (or similar) would communicate its purpose better?

Also, why the default value for the `trucks` arg?  Does it make sense to allow callers to call this without providing that argument?

File: app/speak_slack.rb [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+      msg = ""
+      trucks.uniq!
+      trucks.each do |truck|
+        msg << "\n#{truck}"
+      end
```
Comment (myronmarston) [inline app/speak_slack.rb:11]: This logic relies extensively on mutating things:
- `trucks` is mutated via the `trucks.uniq!` call which means this method both returns a value and changes the array passed by the caller.  That's usually a bad thing; it leads to surprises if the caller does something like this:

``` ruby
trucks = get_trucks_from_somwhere(...)
slack_format = FoodTruck:: SpeakSlack.speak_slack(trucks)
puts "Formatted original #{trucks.inspect} into slack format: #{slack_format}"
```

Here the caller is assuming `speak_slack` is only returning a value and not also mutating the arg (why would it expect the arg to be mutated, after all?).

`msg` is also being mutated throughout.  I'd recommend this instead:

``` ruby
if trucks.empty?
  MSG_NO_FOOD_TRUCKS
else
  trucks.uniq.join("\n")
end
```

File: spec/acceptance/food_truck_api_spec.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+module FoodTruck
+  RSpec.describe 'Food Truck API' do
+    include Rack::Test::Methods
+
+    def app
+      API.new
+    end
+
+    describe 'provides message' do
+      it 'of listed food trucks' do
```
Comment (myronmarston) [inline spec/acceptance/food_truck_api_spec.rb:13]: "it of listed food trucks" reads like broken English.  I'd recommend something like this:

``` ruby
describe 'get /api' do
  it "returns a slack-formatted list of available food trucks" do
    # ...
  end
end
```

"it returns a slack-formatted list of available food trucks" reads much more like understandable English :).

File: app/api.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# Handles web request/response
+require 'sinatra/base'
+require_relative 'food_truck'
+require_relative 'speak_slack'
+
+module FoodTruck
+  class API < Sinatra::Base
+    get '/api' do
```
Comment (myronmarston) [inline app/api.rb:8]: `/api` seems odd for this route.  If/when you add more routes to access more resources, in what sense does it make sense to call this route "api" and not the others?  Should it be called `/available-food-trucks` or `/food-trucks` or something instead?

File: spec/acceptance/food_truck_api_spec.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+        get '/api'
+        expect(last_response.status).to eq(200)
```
Comment (myronmarston) [inline spec/acceptance/food_truck_api_spec.rb:15]: It would be nice to have some stronger expectations here.  For example, if I replace your sinatra route logic with this:

``` ruby
get "/api" do
end
```

...this spec still passes.  I recommend you take a two-part approach to testing this:
- Write a unit spec (similar to the unit spec chapter in the book) that injects a test double in place of the `FoodTruck::SchedulerFetcher` (which I suggested you create above) so that you can have a fast, deterministic unit test with specific, strong expectations about the returned response body.
- Write an acceptance spec that use the real `FoodTruck::SchedulerFetcher` (as a way to ensure everything works together end-to-end when fetching against the real website).  You'll still want some kind of expectations about the response: maybe you can assert something about the general slack format of the output or assert that it returned at least one food truck or something.

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+  config.mock_with :rspec do |mocks|
+    # Prevents you from mocking or stubbing a method that does not exist on
+    # a real object. This is generally recommended, and will default to
+    # `true` in RSpec 4.
+    mocks.verify_partial_doubles = true
+  end
+
+# The settings below are suggested to provide a good initial experience
+# with RSpec, but feel free to customize to your heart's content.
+=begin
```
Comment (myronmarston) [inline spec/spec_helper.rb:46]: I recommend you remove `=begin` and `=end` to get the recommended settings.

File: spec/unit/food_truck_spec.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
+require_relative '../../app/food_truck'
+require_relative '../../constants'
+
+module FoodTruck
+  RSpec.describe FoodTruck do
+    describe 'FoodTruck loads data' do
```
Comment (myronmarston) [inline spec/unit/food_truck_spec.rb:6]: Does this `describe` block serve a useful purpose?  It looks like the example can move into the outer, parent example group.

File: spec/unit/speak_slack_spec.rb [hunk 10]
```diff
@@ -0,0 +1,22 @@
+require_relative '../../app/speak_slack'
+require_relative '../../constants'
+
+module FoodTruck
+  RSpec.describe SpeakSlack do
+    describe 'SpeakSlack says' do
```
Comment (myronmarston) [inline spec/unit/speak_slack_spec.rb:6]: Here too, I'm not sure what purpose this nested example group serves. Notice the documentation output:

```
FoodTruck::SpeakSlack
  SpeakSlack says
    when passed an empty array
      food trucks cannot be found
    when passed an array with food trucks
      lists food trucks in a slack format
      lists uniq food trucks
```

I'm not sure what "SpeakSlack says" communicates that the outer bit ("FoodTruck::SpeakSlack") did not already communicate.

File: spec/unit/speak_slack_spec.rb [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+      context 'when passed an array with food trucks' do
+        it 'lists food trucks in a slack format' do
+          expect(SpeakSlack.trucks(trucks: %w(one two))).to match(/one\ntwo/)
+        end
+        it 'lists uniq food trucks' do
+          expect(SpeakSlack.trucks(trucks: %w(one two two))).to match("one\ntwo")
+        end
+      end
+    end
+  end
```
Comment (myronmarston) [inline spec/unit/speak_slack_spec.rb:21]: The formatting for this works but serves to squish everything together, making it harder to read, IMO.  Most RSpec users (including myself) would format this like so:

``` ruby
module FoodTruck
  RSpec.describe SpeakSlack do
    describe 'SpeakSlack says' do
      context 'when passed an empty array' do
        it 'food trucks cannot be found' do
          expect(SpeakSlack.trucks).to be(MSG_NO_FOOD_TRUCKS)
        end
      end

      context 'when passed an array with food trucks' do
        it 'lists food trucks in a slack format' do
          expect(SpeakSlack.trucks(trucks: %w(one two))).to match(/one\ntwo/)
        end

        it 'lists uniq food trucks' do
          expect(SpeakSlack.trucks(trucks: %w(one two two))).to match("one\ntwo")
        end
      end
    end
  end
end
```

...similar to how most rubyists put blank lines between method definitions.

File: spec/unit/speak_slack_spec.rb [hunk 12]
```diff
@@ -0,0 +1,22 @@
+require_relative '../../app/speak_slack'
+require_relative '../../constants'
+
+module FoodTruck
+  RSpec.describe SpeakSlack do
+    describe 'SpeakSlack says' do
+      context 'when passed an empty array' do
+        it 'food trucks cannot be found' do
```
Comment (myronmarston) [inline spec/unit/speak_slack_spec.rb:8]: "it food trucks cannot be found" is awkward English.  How about `it "returns a string indicating no food trucks can be found"`?

File: spec/unit/speak_slack_spec.rb [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+          expect(SpeakSlack.trucks).to be(MSG_NO_FOOD_TRUCKS)
```
Comment (myronmarston) [inline spec/unit/speak_slack_spec.rb:9]: This expectation is more specific (and therefore more brittle) than needed.  There is no user-observable difference between the method returning the exact `MSG_NO_FOOD_TRUCKS` string instance (which `be(MSG_NO_FOOD_TRUCKS)` requires to pass) and a different string instance containing the same contents -- but if your code returned a different (but equivalent) string instance this spec would fail.  I'd recommend using `eq` instead.

**Comment by myronmarston** [comment]
@TravisSpangle if you plan to keep working on this I recommend you get https://travis-ci.org/ setup for it so that the test suite is run on each push (or each PR) to verify it still passes.  That ensures you don't get in a situation where "it passes for me on my machine" but not for others.  Travis is free for public github projects so there's no real down side to using it.  I use travis on all work projects and all my open source projects.

File: spec/unit/food_truck_spec.rb (hunk 9 above)
Comment (TravisSpangle) [inline spec/unit/food_truck_spec.rb:6]: Is that because ScheduleParser only has one It block or because it's never expected to have behavior other than loading data?

Comment (myronmarston) [inline spec/unit/food_truck_spec.rb:6]: Nesting like this is fine -- it does not cause any harm -- but usually it is done for a reason, and it's not clear to me what you are trying to achieve by nesting it.  This could just be:

``` ruby
module FoodTruck
  RSpec.describe FoodTruck do
    it 'provides food truck names' do
      # ...
    end
  end
end
```

...and it would work just fine.  Your nesting isn't making it behave any different other than changing your spec's full description from the readable `FoodTruck provides food truck names` to the unreadable `FoodTruck FoodTruck loads data provides food truck names`.
